File: fixtures/xlang-cadence/dart/CadenceTrace.dart
// CadenceTrace.dart — PC3 cadence-trace emitter (RFC-0009 §cadence), Dart side.
//
// Emits the packed decision log for the deterministic arrival trace shared
// with the TS/Kotlin/Swift emitters (fixtures/xlang-cadence/):
//
//   state = SEED (default 0x00C0FFEE)
//   for i in 0..N:
//     state = xorshift32(state); arrivals = state % 5; latest += arrivals
//     for policy in [LATEST_WINS, PACED_INTERPOLATE, BURST_COALESCE]:
//       d = policy.step(latest)
//       emit byte1 = (present<<7) | (interp<<6) | (alphaQ12 >> 7)
//       emit byte2 = min(coalesced, 255)
//
// Hex-encoded (lowercase, no separators, one trailing newline) — 6 bytes
// per tick, three policies in kind order. run.sh byte-compares all ports.
//
// Build & run (no pub get — the governor module has zero imports):
//   dart fixtures/xlang-cadence/dart/CadenceTrace.dart [STEPS [SEED]]

import '../../../core/dart/governor.dart';

int _xorshift32(int x0) {
  var x = x0 & 0xFFFFFFFF;
  x = (x ^ ((x << 13) & 0xFFFFFFFF)) & 0xFFFFFFFF;
  x = (x ^ (x >>> 17)) & 0xFFFFFFFF;
  x = (x ^ ((x << 5) & 0xFFFFFFFF)) & 0xFFFFFFFF;
  return x & 0xFFFFFFFF;
}

void main(List<String> args) {
  final steps = args.isNotEmpty ? int.parse(args[0]) : 10000;
  var seed = 0x00C0FFEE;
  if (args.length > 1) {
    seed = args[1].startsWith('0x')
        ? int.parse(args[1].substring(2), radix: 16)
        : int.parse(args[1]);
  }

  final pols = [
    CadencePolicy(CadenceConfig(CadencePolicyKind.latestWins)),
    CadencePolicy(CadenceConfig(CadencePolicyKind.pacedInterpolate)),
    CadencePolicy(CadenceConfig(CadencePolicyKind.burstCoalesce)),
  ];
  final sb = StringBuffer();
  var state = seed;
  var latest = 0;
  for (var i = 0; i < steps; i++) {
    state = _xorshift32(state);
    latest += state % 5;
    for (final p in pols) {
      final a = p.step(latest);
      final b1 = (a.present ? 1 : 0) << 7 |
          (a.interp ? 1 : 0) << 6 |
          (a.alphaQ12 >> 7);
      final b2 = a.coalesced < 255 ? a.coalesced : 255;
      sb.write(b1.toRadixString(16).padLeft(2, '0'));
      sb.write(b2.toRadixString(16).padLeft(2, '0'));
    }
  }
  print(sb.toString());
}

File: fixtures/xlang-cadence/kotlin/CadenceTrace.kt
// CadenceTrace.kt — PC3 cadence-trace emitter (RFC-0009 §cadence), Kotlin side.
//
// Emits the packed decision log for the deterministic arrival trace shared
// with the TS/Swift/Dart emitters (fixtures/xlang-cadence/):
//
//   state = SEED (default 0x00C0FFEE)
//   for i in 0..N:
//     state = xorshift32(state); arrivals = state % 5; latest += arrivals
//     for policy in [LATEST_WINS, PACED_INTERPOLATE, BURST_COALESCE,
//                    PREDICTIVE_PACED]:
//       d = policy.step(latest)
//       emit byte1 = (present<<7) | (interp<<6) | (alphaQ12 >> 7)
//       emit byte2 = min(coalesced, 255)
//
// Hex-encoded (lowercase, no separators, one trailing newline) — 8 bytes
// per tick, four policies in kind order (PC3 v2, RFC-0012 added kind 3).
// run.sh byte-compares all ports.
//
// Build & run:
//   kotlinc core/kotlin/Governor.kt fixtures/xlang-cadence/kotlin/CadenceTrace.kt \
//           -include-runtime -d /tmp/cad-trace-kotlin.jar
//   java -jar /tmp/cad-trace-kotlin.jar [STEPS [SEED]]

package dev.weft.vmtrace

import dev.weft.CadencePolicy
import dev.weft.CadencePolicyKind

private fun xorshift32(x0: Long): Long {
    var x = x0 and 0xffffffffL
    x = x xor ((x shl 13) and 0xffffffffL)
    x = x xor (x ushr 17)
    x = x xor ((x shl 5) and 0xffffffffL)
    return x and 0xffffffffL
}

fun main(args: Array<String>) {
    val steps = if (args.isNotEmpty()) args[0].toInt() else 10_000
    val seed = if (args.size > 1) {
        if (args[1].startsWith("0x")) args[1].substring(2).toLong(16) else args[1].toLong()
    } else 0x00C0FFEEL

    val pols = arrayOf(
        CadencePolicy(CadencePolicyKind.LATEST_WINS),
        CadencePolicy(CadencePolicyKind.PACED_INTERPOLATE),
        CadencePolicy(CadencePolicyKind.BURST_COALESCE),
        CadencePolicy(CadencePolicyKind.PREDICTIVE_PACED)
    )
    val sb = StringBuilder(steps * 6 * 2)
    var state = seed
    var latest = 0L
    for (i in 0 until steps) {
        state = xorshift32(state)
        latest += state % 5
        for (p in pols) {
            val a = p.step(latest)
            val b1 = (((if (a.present) 1 else 0) shl 7) or
                ((if (a.interp) 1 else 0) shl 6) or
                (a.alphaQ12 shr 7)) and 0xff
            val b2 = (minOf(a.coalesced, 255L)).toInt() and 0xff
            sb.append(b1.toString(16).padStart(2, '0'))
            sb.append(b2.toString(16).padStart(2, '0'))
        }
    }
    println(sb.toString())
}
